source/slang/slang-lower-to-ir.cpp

```diff
@@ -6228,6 +6310,79 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         return Slang::maybeGetConstExprType(getBuilder(), type, decl);
     }
 
+        /// Emit appropriate generic parameters for a constraint, and return the value of that constraint.
+        ///
+        /// The `supType` paramete represents the super-type that a parameter is constrained to.
+    IRInst* emitGenericConstraintValue(
+        IRGenContext*               subContext,
+        GenericTypeConstraintDecl*  constraintDecl,
+        IRType*                     supType)
+    {
+        auto subBuilder = subContext->irBuilder;
+
+        // There are two cases we care about here.
+        //
+        if(auto andType = as<IRConjunctionType>(supType))
+        {
+            // The non-trivial case is when the constraint on a generic parameter
+            // was of the form `T : A & B`. In this case, we really want to
+            // emit the function with parameters for each of the two independent
+            // constraints `T : A` and `T : B`.
+            //
+            // We will loop over the "cases" of the conjunction (since
+            // the `IRConunctionType` can support more than just binary
+            // conjunctions) and recursively add constraints for each.
+            //
+            List<IRInst*> caseVals;
+            auto caseCount = andType->getCaseCount();
+            for(Int i = 0; i < caseCount; ++i)
+            {
+                auto caseType = andType->getCaseType(i);
+                auto caseVal = emitGenericConstraintValue(subContext, constraintDecl, caseType);
+                caseVals.add(caseVal);
```

Comment:
So instead of generating a flattened list of constraint values, we are still grouping them into hierarchies of tuples here? Will they get flattened later?

Reply:
It's a little thorny, yeah.

When we are generating code for a `GenericDecl` and it has a `GenericConstraintDecl` that represents something like `T : L & R`, then this code will recursively emit distinct IR parameters for witness tables representing `T : L` and `T : R`. That works reasonably well and ensures that our `IRGeneric`s just have ordinary witness-table parameters and code that works with them doesn't need to know about the tuples.

The problem is that the AST-level code inside of that `GenericDecl` will include references directly to the original `GenericConstraintDecl` (the one for `T : L & R`), and we somehow need to be able to resolve what those references should actually refer to. The decision I made is that given the way a constraint parameter for `T : L & R` is being lowered into the IR as two witness-table parameters for `T : L` and `T : R`, then it made sense for the IR-level value of the original constraint parameter to be represented by the IR tuple `( T:L, T:R )`.

Any piece of AST code that references that `GenericConstraintDecl` is likely to be referencing it via an `ExtractFromConjunctionSubtypeWitness`. For example, if we have a value `value : T` and `T : IFirst & ISecond` and somebody writes `value.getFirst()`, then the front-end will know that it needs to cast `value` to the type `IFirst` to access the `getFirst()` method. In order to do that cast we need a witness that `T : IFirst`, which the AST forms (during lookup) by extracting the left half of the `T : IFirst & ISecond` witness.

Given the encoding of a `GenericConstraintDecl` for `T : L & R` as the tuple `( T:L, T:R )`, it becomes easy to see that an `ExtractFromConjunctionSubtypeWitness` can easily extract the `T:L` part of that constraint by grabbing the first element of the tuple.

Thus, while the lowering of the `GenericConstraintDecl` introduces a few IR tuples, use sites almost always start by extracting one or the other element from it. This PR also includes a change to the `IRBuilder::emitGetTupleElement` logic so that it detects an attempt to emit a `getTupleElement()` on a `makeTuple()` instruction and immediately simplifies it to a direct reference to the appropriate element. As a result, most of the existing IR machinery never has to contend with the possibility of a tuple.

(I originally thought that I might be able to consistently "flatten" these conjunctions everywhere in the IR, but there is a pretty deep assumption in the IR lowering that most AST-level declarations/types/expressions can always lower to a single IR instruction. Yes, there is the `LoweredValInfo` approach, but that really only exists to support l-value vs r-value distinctions during expression lowering (and even *that* might be wearing out its welcome as an approach). In practice, for something like `typealias IBoth = IFirst & ISecond` we fundamentally needed to be able to construct a single `IRInst*` that represents the type being aliased there.)